scipy/cluster/hierarchy.py

```diff
@@ -1395,6 +1400,184 @@ def cut_tree(Z, n_clusters=None, height=None):
+        corresponding cluster id. The cluster id is an integer which is higher
+        for deeper tree levels.
+    cluster_level : ndarray
+        One-dimensional array of integer arrays containing for each input
+        sample its corresponding cluster tree level, i.e. a sequence of
+        0's and 1's. Note that the cluster level is longer for deeper tree
+        levels, being [0] the root cluster, [0, 0] and [0, 1] its offspring,
+        and so on. Also note that in each cluster splitting, the label 0
+        denotes the bigger cluster, while the label 1 denotes the smallest.
+
+    See Also
+    --------
+    cut_tree
+
+    Examples
+    --------
+    >>> from scipy.cluster import hierarchy
+    >>> from scipy import stats
+
+    Initialize the random seed.
+
+    >>> np.random.seed(14)
+
+    Create a input matrix containing 100 data samples with 4 dimensions.
+    Note: using `gamma` in order to generate an unbalanced distribution.
+    If a regular ``cut_tree()`` would be performed, one big and many small
+    clusters would be obtained.
+
+    >>> X = stats.gamma.rvs(0.1, size=400).reshape((100, 4))
+
+    Compute the linkage matrix using the scipy ward() or linkage() method:
+
+    >>> Z = hierarchy.ward(X)
+
+    Perform a balanced cut tree of the linkage matrix:
+
+    >>> cluster_id, cluster_level = hierarchy.cut_tree_balanced(
+    ...                                              Z, max_cluster_size=10)
+    >>> cluster_id[:10]
+    array([19,  4, 10, 12, 20, 12, 14,  9, 15,  2])
+    >>> cluster_level[:10]
+    array([array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 1]), array([0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
+           array([0, 0, 1, 0])], dtype=object)
+
+    """
+    # Assert that the input max_cluster_size is >= 1
+    if not max_cluster_size >= 1:
+        raise ValueError("max_cluster_size should be >= 1, is: {}"
+                         .format(max_cluster_size))
+
+    # Perform a full cut tree of the linkage matrix
+    full_cut = cut_tree(Z)
+
+    # Initialize the variable containing the current cluster id (it will be
+    # higher for each newly found valid cluster)
+    last_cluster_id = 1
+
+    # Initialize the resulting cluster id vector (containing for each row in
+    # input_data_x_sample its corresponding cluster id)
+    cluster_id = np.zeros(full_cut.shape[1], dtype=int)
+
+    # Initialize the resulting cluster level vector (containing for each data
+    # sample its corresponding cluster tree level)
+    cluster_level = np.empty((full_cut.shape[1],), dtype=object)
+    for i in range(full_cut.shape[1]):
```

Comment:
I think the main issue is the following: here we need a `ndarray` storing for each point to be clustered a `ndarray` of variable size, which represents its cluster_level. We should not fill with empty arrays, because at least one level will be there representing one cluster (ID=0) which contains all data points.
Further, I think the general challenge was to create a multidimension numpy array with a varying row size (see the [relevant stachoverflow issue](https://stackoverflow.com/questions/3386259/how-to-make-a-multidimension-numpy-array-with-a-varying-row-size). This was the best way I could find in order to create such a `ndarray`.
Please let me know if you see a more elegant way of initializing such a `ndarray`, so that the method works properly.